src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -2666,32 +2685,40 @@ .maxstack  3
     IL_00fd:  ldloc.3
     IL_00fe:  call       ""void System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool>.SetException(System.Exception)""
     IL_0103:  nop
-    IL_0104:  leave.s    IL_0129
+    IL_0104:  ldarg.0
+    IL_0105:  ldflda     ""System.Runtime.CompilerServices.AsyncIteratorMethodBuilder C.<M>d__0.<>t__builder""
+    IL_010a:  call       ""void System.Runtime.CompilerServices.AsyncIteratorMethodBuilder.Complete()""
+    IL_010f:  nop
+    IL_0110:  leave.s    IL_0141
   }
   // sequence point: }
-  IL_0106:  ldarg.0
-  IL_0107:  ldc.i4.s   -2
-  IL_0109:  stfld      ""int C.<M>d__0.<>1__state""
+  IL_0112:  ldarg.0
+  IL_0113:  ldc.i4.s   -2
+  IL_0115:  stfld      ""int C.<M>d__0.<>1__state""
   // sequence point: <hidden>
-  IL_010e:  ldarg.0
-  IL_010f:  ldflda     ""System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool> C.<M>d__0.<>v__promiseOfValueOrEnd""
-  IL_0114:  ldc.i4.0
-  IL_0115:  call       ""void System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool>.SetResult(bool)""
-  IL_011a:  nop
-  IL_011b:  ret
-  IL_011c:  ldarg.0
-  IL_011d:  ldflda     ""System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool> C.<M>d__0.<>v__promiseOfValueOrEnd""
-  IL_0122:  ldc.i4.1
-  IL_0123:  call       ""void System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool>.SetResult(bool)""
-  IL_0128:  nop
-  IL_0129:  ret
+  IL_011a:  ldarg.0
+  IL_011b:  ldflda     ""System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool> C.<M>d__0.<>v__promiseOfValueOrEnd""
+  IL_0120:  ldc.i4.0
+  IL_0121:  call       ""void System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool>.SetResult(bool)""
+  IL_0126:  nop
+  IL_0127:  ldarg.0
+  IL_0128:  ldflda     ""System.Runtime.CompilerServices.AsyncIteratorMethodBuilder C.<M>d__0.<>t__builder""
+  IL_012d:  call       ""void System.Runtime.CompilerServices.AsyncIteratorMethodBuilder.Complete()""
+  IL_0132:  nop
+  IL_0133:  ret
+  IL_0134:  ldarg.0
+  IL_0135:  ldflda     ""System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool> C.<M>d__0.<>v__promiseOfValueOrEnd""
+  IL_013a:  ldc.i4.1
+  IL_013b:  call       ""void System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool>.SetResult(bool)""
```

Comment:
Why don't we need to call `AsyncIteratorMethodBuilder.Complete` here?

Reply:
The tail of the `MoveNext()` method is like this:
``` 
            // ... _exprReturnLabel: ...
            // ... this.state = FinishedState; ...
            // if (this.combinedTokens != null) { this.combinedTokens.Dispose(); this.combinedTokens = null; } // for enumerables only
            // this.promiseOfValueOrEnd.SetResult(false);
            // this.builder.Complete();
            // return;

            // _exprReturnLabelTrue:
            // this.promiseOfValueOrEnd.SetResult(true);

            // ... _exitLabel: ...
            // ... return; ...
```

It has different ways of returning from `MoveNext()` (in non-exceptional flow): 
- one for signaling that we have no values left: we set the finished state, dispose our `CancellationTokenSource`, record that we're done with `SetResult(false)` and dispose the builder with `Complete()`.
- the other for returning a value in the middle of the stream: we have already save that value in `Current`, we just signal with `SetResult(true)`. We don't need to dispose any resources, since we expect to process more values in the stream.

---
In reply to: [337657566](https://github.com/dotnet/roslyn/pull/39436#discussion_r337657566) [](ancestors = 337657566)

Reply:
Also it helps to look at the example I linked to in description showing current tail of `MoveNext()`. Only two of the exit paths need to dispose.
```
            }
            catch (Exception exception)
            {
                <>1__state = -2;
                <>v__promiseOfValueOrEnd.SetException(exception);
                return;
            }
            <>1__state = -2;
            <>v__promiseOfValueOrEnd.SetResult(false);
            return;
            IL_00ed:
            <>v__promiseOfValueOrEnd.SetResult(true);
        }
```

---
In reply to: [337681134](https://github.com/dotnet/roslyn/pull/39436#discussion_r337681134) [](ancestors = 337681134,337657566)

Reply:
Ah, I see, thanks